// @microsoft/iot-cardboard-js/internal/AssetSimulation-f7449f98.js
import { C as ComponentError, be as __awaiter, bf as __generator, bE as __rest, d as ComponentErrorType, bg as __assign, H as HierarchyNodeType, bj as __spreadArray, bF as downloadJSON, I as AdtPusherSimulationType } from './Interfaces-f1972c79.js';
import { a as axios } from './index-f67cd73f.js';
import { A as AdapterResult } from './AdapterResult-96d676e4.js';

var ADTAdapterModelsData = /** @class */ (function () {
    function ADTAdapterModelsData(data) {
        this.data = data;
    }
    ADTAdapterModelsData.prototype.hasNoData = function () {
        var _a;
        return (this.data === undefined ||
            this.data === null ||
            ((_a = this.data.value) === null || _a === void 0 ? void 0 : _a.length) === 0);
    };
    return ADTAdapterModelsData;
}());
var ADTAdapterTwinsData = /** @class */ (function () {
    function ADTAdapterTwinsData(data) {
        this.data = data;
    }
    ADTAdapterTwinsData.prototype.hasNoData = function () {
        return (this.data === undefined ||
            this.data === null ||
            this.data.value.length === 0);
    };
    return ADTAdapterTwinsData;
}());
var ADTAdapterSearchByQueryData = /** @class */ (function () {
    function ADTAdapterSearchByQueryData(data) {
        this.data = data;
    }
    ADTAdapterSearchByQueryData.prototype.hasNoData = function () {
        return (this.data === undefined ||
            this.data === null ||
            this.data.value.length === 0);
    };
    return ADTAdapterSearchByQueryData;
}());
var ADTAdapterPatchData = /** @class */ (function () {
    function ADTAdapterPatchData(data) {
        this.data = data;
    }
    ADTAdapterPatchData.prototype.hasNoData = function () {
        return this.data === null || this.data === undefined;
    };
    return ADTAdapterPatchData;
}());

var ADTModelData = /** @class */ (function () {
    function ADTModelData(data) {
        this.data = data;
    }
    ADTModelData.prototype.hasNoData = function () {
        return this.data === null || this.data === undefined;
    };
    return ADTModelData;
}());
var ADTAllModelsData = /** @class */ (function () {
    function ADTAllModelsData(data) {
        this.data = data;
    }
    ADTAllModelsData.prototype.hasNoData = function () {
        var _a, _b;
        return !((_a = this.data) === null || _a === void 0 ? void 0 : _a.parsedModels) || !((_b = this.data) === null || _b === void 0 ? void 0 : _b.rawModels);
    };
    return ADTAllModelsData;
}());
var ADTTwinToModelMappingData = /** @class */ (function () {
    function ADTTwinToModelMappingData(data) {
        this.data = data;
    }
    ADTTwinToModelMappingData.prototype.hasNoData = function () {
        var _a, _b;
        return !((_a = this.data) === null || _a === void 0 ? void 0 : _a.twinId) || !((_b = this.data) === null || _b === void 0 ? void 0 : _b.modelId);
    };
    return ADTTwinToModelMappingData;
}());

var ADTRelationshipsData = /** @class */ (function () {
    function ADTRelationshipsData(data) {
        this.data = data;
    }
    ADTRelationshipsData.prototype.hasNoData = function () {
        return this.data === null || this.data.length === 0;
    };
    return ADTRelationshipsData;
}());
var ADTRelationshipData = /** @class */ (function () {
    function ADTRelationshipData(data) {
        this.data = data;
    }
    ADTRelationshipData.prototype.hasNoData = function () {
        return this.data === null || this.data.length === 0;
    };
    return ADTRelationshipData;
}());

var ADTTwinData = /** @class */ (function () {
    function ADTTwinData(data) {
        this.data = data;
    }
    ADTTwinData.prototype.hasNoData = function () {
        return this.data === null || this.data === undefined;
    };
    return ADTTwinData;
}());

var KeyValuePairAdapterData = /** @class */ (function () {
    function KeyValuePairAdapterData(data) {
        this.data = data;
    }
    KeyValuePairAdapterData.prototype.hasNoData = function () {
        return (this.data === null ||
            this.data.length === 0 ||
            !this.data.reduce(function (acc, curr) {
                return acc || curr.value;
            }, false));
    };
    return KeyValuePairAdapterData;
}());

/** Utility class which creates sandbox environment for adapter data fetching.
 *
 * • Manages errors and catasrophicErrors.
 *
 * • Safely fetches auth token and passes token to data fetching callback
 *
 * • Catches, classifies, and aggregates errors
 */
var AdapterMethodSandbox = /** @class */ (function () {
    function AdapterMethodSandbox(authservice) {
        this.errors = [];
        this.catasrophicError = null;
        this.authService = authservice;
    }
    /**
     *  Pushes new ComponentError onto errors list.  If error is marked as catastrophic,
     *  execution will halt with catastrophic error attached to result
     */
    AdapterMethodSandbox.prototype.pushError = function (_a) {
        var rawError = _a.rawError, message = _a.message, type = _a.type, isCatastrophic = _a.isCatastrophic;
        var error = new ComponentError({
            message: message,
            type: type,
            isCatastrophic: isCatastrophic,
            rawError: rawError
        });
        this.errors.push(error);
        if (error.isCatastrophic) {
            this.catasrophicError = error;
            throw error;
        }
    };
    /**
     * Fetch token wrapped in try / catch block.  If token fetch fails, will attach
     * catastrophic TokenRetrievalFailed error, halting further execution.
     */
    AdapterMethodSandbox.prototype.safelyFetchToken = function (tokenFor) {
        return __awaiter(this, void 0, void 0, function () {
            var token, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // If adapterMethodSandbox not constructed with authService, skip token fetching
                        if (!this.authService) {
                            return [2 /*return*/, null];
                        }
                        token = '';
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.authService.getToken(tokenFor)];
                    case 2:
                        token = _a.sent();
                        if (!token && token !== '') {
                            throw new Error('Token undefined');
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        err_1 = _a.sent();
                        this.pushError({
                            isCatastrophic: true,
                            type: ComponentErrorType.TokenRetrievalFailed,
                            rawError: err_1
                        });
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, token];
                }
            });
        });
    };
    /**
     * Wraps adapter data-fetching logic in sandbox which safely fetches an auth token and handles errors consistently.
     *
     * • Any operation that may throw in the callback passed to this method should be wrapped in additional try/catch
     * blocks which push specific error info to the sandbox.
     *
     * • Errors marked isCatastrophic will halt adapter execution.
     *
     * • Uncaught errors will be treated as catastrophic errors with unknown types and will also halt adapter execution.
     *  */
    AdapterMethodSandbox.prototype.safelyFetchData = function (fetchDataWithToken, tokenFor) {
        return __awaiter(this, void 0, void 0, function () {
            var token, data, err_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        return [4 /*yield*/, this.safelyFetchToken(tokenFor)];
                    case 1:
                        token = _a.sent();
                        return [4 /*yield*/, fetchDataWithToken(token)];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, new AdapterResult({
                                errorInfo: {
                                    errors: this.errors,
                                    catastrophicError: this.catasrophicError
                                },
                                result: data
                            })];
                    case 3:
                        err_2 = _a.sent();
                        // Uncaught errors are bubbled up and caught here.
                        if (!(err_2 instanceof ComponentError)) {
                            // Unknown error, construct new catastrophicError error
                            this.catasrophicError = new ComponentError({
                                isCatastrophic: true,
                                rawError: err_2,
                                type: ComponentErrorType.UnknownError
                            });
                            this.errors.unshift(this.catasrophicError);
                        }
                        else if (!this.catasrophicError) {
                            // Uncaught ComponentError thrown explicitly (not pushed to sandbox).  Attach to catastrophicError.
                            this.catasrophicError = err_2;
                            this.errors.unshift(this.catasrophicError);
                        }
                        // Attach errorInfo, nullify result, and return AdapterResult.
                        return [2 /*return*/, new AdapterResult({
                                errorInfo: {
                                    errors: this.errors,
                                    catastrophicError: this.catasrophicError
                                },
                                result: null
                            })];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    AdapterMethodSandbox.prototype.safelyFetchDataCancellableAxiosPromise = function (returnDataClass, axiosParams, dataTransformFunc, tokenFor) {
        var _this = this;
        var headers = axiosParams.headers, restOfParams = __rest(axiosParams, ["headers"]);
        var cancelTokenSource = axios.CancelToken.source();
        var cancellablePromise = this.safelyFetchData(function (token) { return __awaiter(_this, void 0, void 0, function () {
            var axiosData, err_3, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, axios(__assign(__assign({}, restOfParams), { headers: __assign({ 'Content-Type': 'application/json', authorization: 'Bearer ' + token }, headers), cancelToken: cancelTokenSource.token }))];
                    case 1:
                        axiosData = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        err_3 = _a.sent();
                        if (axios.isCancel(err_3)) {
                            this.pushError({
                                type: ComponentErrorType.DataFetchFailed,
                                isCatastrophic: false,
                                rawError: err_3
                            });
                        }
                        else {
                            this.pushError({
                                type: ComponentErrorType.DataFetchFailed,
                                isCatastrophic: true,
                                rawError: err_3
                            });
                        }
                        return [3 /*break*/, 3];
                    case 3:
                        result = axiosData === null || axiosData === void 0 ? void 0 : axiosData.data;
                        return [2 /*return*/, new returnDataClass(dataTransformFunc ? dataTransformFunc(result) : result)];
                }
            });
        }); }, tokenFor);
        cancellablePromise.cancel = cancelTokenSource.cancel;
        return cancellablePromise;
    };
    return AdapterMethodSandbox;
}());

var HierarchyNode = /** @class */ (function () {
    function HierarchyNode() {
    }
    HierarchyNode.createNodesFromADTModels = function (models, nodeType, newlyAddedModelIds) {
        if (nodeType === void 0) { nodeType = HierarchyNodeType.Parent; }
        return models
            ? models
                .sort(function (a, b) {
                var _a, _b;
                return (((_a = a.displayName) === null || _a === void 0 ? void 0 : _a.en) || a.id).localeCompare(((_b = b.displayName) === null || _b === void 0 ? void 0 : _b.en) || b.id, undefined, {
                    numeric: true,
                    sensitivity: 'base'
                });
            })
                .reduce(function (p, c) {
                var _a;
                p[c.id] = __assign(__assign({ name: ((_a = c.displayName) === null || _a === void 0 ? void 0 : _a.en) || c.id, id: c.id, nodeData: c, nodeType: nodeType }, (nodeType === HierarchyNodeType.Parent && {
                    children: {},
                    isCollapsed: true
                })), { isNewlyAdded: newlyAddedModelIds === null || newlyAddedModelIds === void 0 ? void 0 : newlyAddedModelIds.includes(c.id) });
                return p;
            }, {})
            : {};
    };
    HierarchyNode.createNodesFromADTTwins = function (twins, modelNode) {
        return twins
            ? twins
                .sort(function (a, b) {
                return a.$dtId.localeCompare(b.$dtId, undefined, {
                    numeric: true,
                    sensitivity: 'base'
                });
            })
                .reduce(function (p, c) {
                p[c.$dtId] = {
                    name: c.$dtId,
                    id: c.$dtId,
                    parentNode: modelNode,
                    nodeData: c,
                    nodeType: HierarchyNodeType.Child
                };
                return p;
            }, {})
            : {};
    };
    return HierarchyNode;
}());

var Asset = /** @class */ (function () {
    function Asset(name) {
        var _this = this;
        this.getDoubleValue = function (minValue, maxValue) {
            return function (currentValue) {
                var direction = currentValue > maxValue
                    ? -1
                    : currentValue < minValue
                        ? 1
                        : Math.random() < 0.5
                            ? -1
                            : 1;
                var step = direction * (Math.random() * (maxValue - minValue) * 0.02);
                return (currentValue += step);
            };
        };
        this.getIntegerValue = function (minValue, maxValue) {
            return function (currentValue) {
                var direction = currentValue >= maxValue
                    ? -1
                    : currentValue <= minValue
                        ? 1
                        : Math.random() < 0.5
                            ? -1
                            : 1;
                return (currentValue += direction);
            };
        };
        this.getBooleanValue = function (isTrueThreshold) {
            return function (_currentValue) {
                return Math.random() > isTrueThreshold;
            };
        };
        this.getStringValue = function () {
            return function (_currentValue) {
                var fourDigitNumber = Math.floor(Math.random() * 1000);
                return "Box" + fourDigitNumber;
            };
        };
        this.properties = [];
        this.relationships = [];
        this.twins = [];
        this.name = name;
        switch (name) {
            case 'RobotArm': {
                this.properties = [
                    {
                        id: this.name,
                        propertyName: 'FailedPickupsLastHr',
                        currentValue: 1,
                        getNextValue: this.getIntegerValue(0, 5),
                        schema: 'integer'
                    },
                    {
                        id: this.name,
                        propertyName: 'PickupFailedAlert',
                        currentValue: false,
                        getNextValue: this.getBooleanValue(0.75),
                        schema: 'boolean'
                    },
                    {
                        id: this.name,
                        propertyName: 'PickupFailedBoxID',
                        currentValue: 'Box1',
                        getNextValue: this.getStringValue(),
                        schema: 'string'
                    },
                    {
                        id: this.name,
                        propertyName: 'HydraulicPressure',
                        currentValue: 20,
                        getNextValue: this.getDoubleValue(10, 100)
                    }
                ];
                [1, 2, 3, 4, 5, 6].forEach(function (idx) {
                    _this.twins.push({
                        name: "Arm" + idx,
                        properties: _this.getAssetProperties()
                    });
                });
                break;
            }
            case 'DistributionCenter': {
                this.relationships.push({
                    name: 'contains',
                    target: 'RobotArm'
                });
                this.twins.push({
                    name: 'DistCtr',
                    assetRelationships: [1, 2, 3, 4, 5, 6].map(function (idx) {
                        return {
                            name: 'contains',
                            target: "Arm" + idx,
                            targetModel: 'RobotArm'
                        };
                    }),
                    properties: []
                });
                break;
            }
            case 'Car': {
                this.properties = [
                    {
                        id: this.name,
                        propertyName: 'Speed',
                        currentValue: Math.floor(Math.random() * 20) + 40,
                        getNextValue: this.getDoubleValue(0, 100)
                    },
                    {
                        id: this.name,
                        propertyName: 'OutdoorTemperature',
                        currentValue: Math.floor(Math.random()) + 40,
                        getNextValue: this.getDoubleValue(20, 80)
                    },
                    {
                        id: this.name,
                        propertyName: 'OilPressure',
                        currentValue: Math.floor(Math.random()) + 30,
                        getNextValue: this.getDoubleValue(28, 32)
                    }
                ];
                this.twins.push({
                    name: 'CarTwin',
                    properties: this.getAssetProperties()
                });
                break;
            }
            case 'Windmill': {
                this.properties = [
                    {
                        id: this.name,
                        propertyName: 'OutdoorTemperature',
                        currentValue: Math.floor(Math.random() * 20) + 40,
                        getNextValue: this.getDoubleValue(0, 100)
                    },
                    {
                        id: this.name,
                        propertyName: 'AtmosphericPressure',
                        currentValue: Math.floor(Math.random()) + 30,
                        getNextValue: this.getDoubleValue(29, 31)
                    },
                    {
                        id: this.name,
                        propertyName: 'WindVelocity',
                        currentValue: Math.floor(Math.random() * 30),
                        getNextValue: this.getDoubleValue(0, 70)
                    },
                    {
                        id: this.name,
                        propertyName: 'BearingTemperature',
                        currentValue: Math.floor(Math.random() * 30) + 90,
                        getNextValue: this.getDoubleValue(90, 200)
                    },
                    {
                        id: this.name,
                        propertyName: 'OilViscosity',
                        currentValue: Math.floor(Math.random() * 5) + 10,
                        getNextValue: this.getDoubleValue(10, 80)
                    }
                ];
                this.twins.push({
                    name: 'Windmill_1',
                    properties: this.getAssetProperties()
                });
                break;
            }
            case 'HVACSystem': {
                this.properties = [
                    {
                        id: this.name,
                        propertyName: 'FanSpeed',
                        currentValue: Math.floor(Math.random() * 20) + 40,
                        getNextValue: this.getDoubleValue(0, 100)
                    },
                    {
                        id: this.name,
                        propertyName: 'CoolerTemperature',
                        currentValue: Math.floor(Math.random()) + 40,
                        getNextValue: this.getDoubleValue(20, 60)
                    },
                    {
                        id: this.name,
                        propertyName: 'HeaterTemperature',
                        currentValue: Math.floor(Math.random()) + 50,
                        getNextValue: this.getDoubleValue(40, 100)
                    }
                ];
                this.twins.push({
                    name: 'HVACSystem_1',
                    properties: this.getAssetProperties()
                });
                break;
            }
            case 'PasteurizationMachine': {
                this.properties = [
                    {
                        id: this.name,
                        propertyName: 'InFlow',
                        currentValue: Math.floor(Math.random() * 300) + 50,
                        getNextValue: this.getDoubleValue(50, 600)
                    },
                    {
                        id: this.name,
                        propertyName: 'OutFlow',
                        currentValue: Math.floor(Math.random() * 300) + 50,
                        getNextValue: this.getDoubleValue(50, 600)
                    },
                    {
                        id: this.name,
                        propertyName: 'Temperature',
                        currentValue: Math.floor(Math.random()) + 120,
                        getNextValue: this.getDoubleValue(110, 250)
                    },
                    {
                        id: this.name,
                        propertyName: 'PercentFull',
                        currentValue: Math.floor(Math.random()),
                        getNextValue: this.getDoubleValue(0, 1)
                    }
                ];
                this.relationships.push({
                    name: 'feeds',
                    target: 'SaltMachine'
                });
                this.twins.push({
                    name: 'PasteurizationMachine_A01',
                    assetRelationships: [
                        {
                            name: 'feeds',
                            target: 'SaltMachine_C0',
                            targetModel: 'SaltMachine'
                        }
                    ],
                    properties: this.getAssetProperties()
                });
                this.twins.push({
                    name: 'PasteurizationMachine_A02',
                    assetRelationships: [
                        {
                            name: 'feeds',
                            target: 'SaltMachine_C0',
                            targetModel: 'SaltMachine'
                        }
                    ],
                    properties: this.getAssetProperties()
                });
                this.twins.push({
                    name: 'PasteurizationMachine_A03',
                    assetRelationships: [
                        {
                            name: 'feeds',
                            target: 'SaltMachine_C1',
                            targetModel: 'SaltMachine'
                        }
                    ],
                    properties: this.getAssetProperties()
                });
                this.twins.push({
                    name: 'PasteurizationMachine_A04',
                    assetRelationships: [
                        {
                            name: 'feeds',
                            target: 'SaltMachine_C2',
                            targetModel: 'SaltMachine'
                        }
                    ],
                    properties: this.getAssetProperties()
                });
                break;
            }
            case 'SaltMachine': {
                this.properties = [
                    {
                        id: this.name,
                        propertyName: 'InFlow',
                        currentValue: Math.floor(Math.random() * 300) + 50,
                        getNextValue: this.getDoubleValue(50, 600)
                    },
                    {
                        id: this.name,
                        propertyName: 'OutFlow',
                        currentValue: Math.floor(Math.random() * 300) + 50,
                        getNextValue: this.getDoubleValue(50, 600)
                    }
                ];
                this.twins.push({
                    name: 'SaltMachine_C0',
                    properties: this.getAssetProperties()
                });
                this.twins.push({
                    name: 'SaltMachine_C1',
                    properties: this.getAssetProperties()
                });
                this.twins.push({
                    name: 'SaltMachine_C2',
                    properties: this.getAssetProperties()
                });
                break;
            }
            case 'MaintenancePersonnel': {
                this.relationships.push({ name: 'maintains' });
                this.twins.push({
                    name: 'Xenia',
                    assetRelationships: [
                        {
                            name: 'maintains',
                            target: 'SaltMachine_C0',
                            targetModel: 'SaltMachine'
                        },
                        {
                            name: 'maintains',
                            target: 'SaltMachine_C1',
                            targetModel: 'SaltMachine'
                        }
                    ],
                    properties: []
                });
                this.twins.push({
                    name: 'Amy',
                    assetRelationships: [
                        {
                            name: 'maintains',
                            target: 'SaltMachine_C1',
                            targetModel: 'SaltMachine'
                        },
                        {
                            name: 'maintains',
                            target: 'PasteurizationMachine_A01',
                            targetModel: 'PasteurizationMachine'
                        }
                    ],
                    properties: []
                });
                this.twins.push({
                    name: 'John',
                    assetRelationships: [
                        {
                            name: 'maintains',
                            target: 'PasteurizationMachine_A02',
                            targetModel: 'PasteurizationMachine'
                        },
                        {
                            name: 'maintains',
                            target: 'PasteurizationMachine_A03',
                            targetModel: 'PasteurizationMachine'
                        }
                    ],
                    properties: []
                });
                this.twins.push({
                    name: 'Phillip',
                    assetRelationships: [
                        {
                            name: 'maintains',
                            target: 'SaltMachine_C2',
                            targetModel: 'SaltMachine'
                        },
                        {
                            name: 'maintains',
                            target: 'PasteurizationMachine_A04',
                            targetModel: 'PasteurizationMachine'
                        }
                    ],
                    properties: []
                });
                break;
            }
            case 'Factory': {
                this.relationships.push({ name: 'contains' });
                this.relationships.push({
                    name: 'employs',
                    target: 'MaintenancePersonnel'
                });
                this.twins.push({
                    name: 'OsloFactory',
                    assetRelationships: [
                        {
                            name: 'contains',
                            target: 'SaltMachine_C0',
                            targetModel: 'SaltMachine'
                        },
                        {
                            name: 'contains',
                            target: 'SaltMachine_C1',
                            targetModel: 'SaltMachine'
                        },
                        {
                            name: 'contains',
                            target: 'PasteurizationMachine_A01',
                            targetModel: 'PasteurizationMachine'
                        },
                        {
                            name: 'contains',
                            target: 'PasteurizationMachine_A02',
                            targetModel: 'PasteurizationMachine'
                        },
                        {
                            name: 'contains',
                            target: 'PasteurizationMachine_A03',
                            targetModel: 'PasteurizationMachine'
                        },
                        {
                            name: 'employs',
                            target: 'Amy',
                            targetModel: 'MaintenancePersonnel'
                        },
                        {
                            name: 'employs',
                            target: 'John',
                            targetModel: 'MaintenancePersonnel'
                        },
                        {
                            name: 'employs',
                            target: 'Xenia',
                            targetModel: 'MaintenancePersonnel'
                        }
                    ],
                    properties: []
                });
                this.twins.push({
                    name: 'StockholmFactory',
                    assetRelationships: [
                        {
                            name: 'contains',
                            target: 'SaltMachine_C2',
                            targetModel: 'SaltMachine'
                        },
                        {
                            name: 'contains',
                            target: 'PasteurizationMachine_A04',
                            targetModel: 'PasteurizationMachine'
                        },
                        {
                            name: 'employs',
                            target: 'Phillip',
                            targetModel: 'MaintenancePersonnel'
                        }
                    ],
                    properties: []
                });
                break;
            }
            case 'Country': {
                this.relationships.push({
                    name: 'contains',
                    target: 'Factory'
                });
                this.twins.push({
                    name: 'Norway',
                    assetRelationships: [
                        {
                            name: 'contains',
                            target: 'OsloFactory',
                            targetModel: 'Factory'
                        }
                    ],
                    properties: []
                });
                this.twins.push({
                    name: 'Sweden',
                    assetRelationships: [
                        {
                            name: 'contains',
                            target: 'StockholmFactory',
                            targetModel: 'Factory'
                        }
                    ],
                    properties: []
                });
                break;
            }
        }
    }
    Asset.prototype.getAssetProperties = function () {
        var assetProperties = [];
        this.properties.forEach(function (property) {
            assetProperties.push(new AssetProperty(property));
        });
        return assetProperties;
    };
    return Asset;
}());
var AssetProperty = /** @class */ (function () {
    function AssetProperty(_a) {
        var id = _a.id, propertyName = _a.propertyName, currentValue = _a.currentValue, getNextValue = _a.getNextValue, _b = _a.schema, schema = _b === void 0 ? 'double' : _b;
        this.id = id;
        this.propertyName = propertyName;
        this.currentValue = currentValue;
        this.getNextValue = getNextValue;
        this.schema = schema;
    }
    AssetProperty.prototype.tick = function () {
        this.currentValue = this.getNextValue(this.currentValue);
        var event = {
            op: 'replace',
            path: "/" + this.propertyName,
            value: this.currentValue
        };
        return event;
    };
    return AssetProperty;
}());

var modelTwinsRelationshipsData = {
    versionNumber: 1
};
var AssetSimulation = /** @class */ (function () {
    function AssetSimulation(seedTimeMillis, intervalMillis, simulationType) {
        this.assets = [];
        this.seedTimeMillis = seedTimeMillis;
        this.intervalMillis = intervalMillis;
        if (simulationType === AdtPusherSimulationType.DairyProduction) {
            this.assets.push(new Asset('PasteurizationMachine'));
            this.assets.push(new Asset('SaltMachine'));
            this.assets.push(new Asset('MaintenancePersonnel'));
            this.assets.push(new Asset('Factory'));
            this.assets.push(new Asset('Country'));
        }
        else if (simulationType === AdtPusherSimulationType.RobotArms) {
            this.assets.push(new Asset('RobotArm'));
            this.assets.push(new Asset('DistributionCenter'));
        }
    }
    AssetSimulation.prototype.generateTwinId = function (name) {
        return name + "Twin";
    };
    AssetSimulation.prototype.generateModelId = function (name) {
        return "dtmi:assetGen:" + name + ";" + modelTwinsRelationshipsData.versionNumber;
    };
    AssetSimulation.prototype.tick = function () {
        this.seedTimeMillis += this.intervalMillis;
        var events = [];
        this.assets.forEach(function (asset) {
            asset.twins.forEach(function (twin) {
                var updateTwin = {
                    dtId: twin.name,
                    patchJSON: twin.properties.map(function (d) {
                        return d.tick();
                    })
                };
                events.push(updateTwin);
            });
        });
        return events;
    };
    AssetSimulation.prototype.generateDTModels = function (download) {
        var _this = this;
        if (download === void 0) { download = false; }
        var dtdlModels = this.assets.map(function (asset) {
            var propertyContents = asset.properties.map(function (property) { return ({
                '@type': 'Property',
                name: property.propertyName,
                schema: property.schema || 'double'
            }); });
            var relationshipContents = asset.relationships.map(function (assetRelationship) {
                var relationship = {
                    '@type': 'Relationship',
                    name: assetRelationship.name,
                    properties: [
                        {
                            '@type': 'Property',
                            name: 'targetModel',
                            schema: 'string'
                        }
                    ]
                };
                if (assetRelationship.target) {
                    relationship.target = _this.generateModelId(assetRelationship.target);
                }
                return relationship;
            });
            var model = {
                '@id': _this.generateModelId(asset.name),
                '@type': 'Interface',
                '@context': 'dtmi:dtdl:context;2',
                displayName: asset.name,
                contents: __spreadArray(__spreadArray([], propertyContents), relationshipContents)
            };
            return model;
        });
        if (download) {
            downloadJSON(JSON.stringify(dtdlModels), 'DT_Models.json');
        }
        return dtdlModels;
    };
    AssetSimulation.prototype.generateDTwins = function (download) {
        if (download === void 0) { download = false; }
        var twins = [];
        this.assets.forEach(function (asset) {
            asset.twins.forEach(function (assetTwin) {
                var twin = {
                    $dtId: assetTwin.name,
                    $metadata: {
                        $model: "dtmi:assetGen:" + asset.name + ";" + modelTwinsRelationshipsData.versionNumber
                    }
                };
                asset.properties.forEach(function (property) {
                    twin["" + property.propertyName] = property.currentValue;
                });
                twins.push(twin);
            });
        });
        if (download) {
            downloadJSON(JSON.stringify(twins), 'DT_Twins.json');
        }
        return twins;
    };
    AssetSimulation.prototype.generateTwinRelationships = function () {
        var relationships = [];
        this.assets.forEach(function (asset) {
            asset === null || asset === void 0 ? void 0 : asset.twins.forEach(function (twin) {
                var _a;
                (_a = twin.assetRelationships) === null || _a === void 0 ? void 0 : _a.forEach(function (relationship, relationshipIndex) {
                    relationships.push({
                        $relId: twin.name + "_" + relationship.name + "_Relationship" + relationshipIndex,
                        $dtId: twin.name,
                        $name: relationship.name,
                        $targetId: relationship.target
                            ? relationship.target
                            : '',
                        targetModel: relationship.targetModel
                            ? relationship.targetModel
                            : ''
                    });
                });
            });
        });
        return relationships;
    };
    return AssetSimulation;
}());

export { ADTAdapterModelsData as A, HierarchyNode as H, KeyValuePairAdapterData as K, ADTAdapterTwinsData as a, ADTModelData as b, ADTRelationshipsData as c, ADTRelationshipData as d, ADTTwinData as e, AdapterMethodSandbox as f, AssetSimulation as g, ADTAdapterSearchByQueryData as h, ADTAdapterPatchData as i, ADTTwinToModelMappingData as j, ADTAllModelsData as k };
