// azure-iot-dtdl-parser/dist/parser/enumInfoImpl.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
/* eslint-disable valid-jsdoc */
/* eslint-disable guard-for-in */
/* eslint-disable no-empty */
/* eslint-disable no-empty-function */
/* eslint-disable no-unused-vars */
/* eslint-disable sort-imports */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumInfoImpl = void 0;
const parserCollection_1 = require("./parserCollection");
const internalDtmi_1 = require("./internalDtmi");
const parsingErrorImpl_1 = require("./parsingErrorImpl");
const aggregateContext_1 = require("./aggregateContext");
const enum_1 = require("./enum");
class EnumInfoImpl {
    constructor(dtdlVersion, id, childOf, definedIn, entityKind) {
        this._enumValuesValueConstraints = [];
        this._enumValuesInstanceProperties = [];
        this._enumValuesAllowedVersionsV2 = new Set().add(2);
        this._enumValuesAllowedVersionsV3 = new Set().add(3);
        this._valueSchemaValueConstraints = [];
        this._valueSchemaInstanceProperties = [];
        this._valueSchemaAllowedVersionsV2 = new Set().add(2);
        this._valueSchemaAllowedVersionsV3 = new Set().add(3);
        this.parserClass = parserCollection_1.ParserCollection.EnumInfoParser;
        this.dtdlVersion = dtdlVersion;
        this.id = id;
        this.childOf = childOf;
        this.definedIn = definedIn;
        this.entityKind = entityKind;
        this.description = {};
        this.displayName = {};
        this.enumValues = [];
        this.supplementalTypeIds = [];
        this.supplementalProperties = {};
        this.supplementalTypes = [];
        this.isPartition = false;
        this.undefinedTypes = [];
        this.undefinedProperties = {};
        this._checkedForDescendantSchemaOrContentsComponentNarrow = false;
        this._idOfDescendantSchemaOrContentsComponentNarrow = undefined;
        this._checkedDescendantEnumValueDatatype = undefined;
        this._checkedForDescendantSchemaArray = false;
        this._idOfDescendantSchemaArray = undefined;
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.NotStarted;
        this._countOfExtendsNarrowValue = 0;
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.NotStarted;
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue = 0;
    }
    static initialize() {
        this._versionlessTypes = new Set()
            .add("dtmi:dtdl:class:ComplexSchema")
            .add("dtmi:dtdl:class:Entity")
            .add("dtmi:dtdl:class:Enum")
            .add("dtmi:dtdl:class:Schema");
    }
    addType(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    dtmi, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    supplementalType) {
        this.supplementalTypeIds.push(dtmi);
        if (supplementalType !== undefined) {
            this.supplementalTypes.push(supplementalType);
        }
        supplementalType.attachConstraints(this);
        supplementalType.bindInstanceProperties(this);
    }
    doesHaveType(typeId) {
        return (EnumInfoImpl._versionlessTypes.has(new internalDtmi_1.InDTMI(typeId).versionless) ||
            this.supplementalTypes.some((st) => st.doesHaveType(typeId)));
    }
    addConstraint(propertyName, valueConstraint) {
        switch (propertyName) {
            case "enumValues":
                if (this._enumValuesValueConstraints === undefined) {
                    this._enumValuesValueConstraints = [];
                }
                this._enumValuesValueConstraints.push(valueConstraint);
                break;
            case "valueSchema":
                if (this._valueSchemaValueConstraints === undefined) {
                    this._valueSchemaValueConstraints = [];
                }
                this._valueSchemaValueConstraints.push(valueConstraint);
                break;
        }
    }
    addInstanceProperty(propertyName, instancePropertyName) {
        switch (propertyName) {
            case "enumValues":
                if (this._enumValuesInstanceProperties === undefined) {
                    this._enumValuesInstanceProperties = [];
                }
                this._enumValuesInstanceProperties.push(instancePropertyName);
                break;
            case "valueSchema":
                if (this._valueSchemaInstanceProperties === undefined) {
                    this._valueSchemaInstanceProperties = [];
                }
                this._valueSchemaInstanceProperties.push(instancePropertyName);
                break;
        }
    }
    doesPropertyDictContainKey(propertyName, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key) {
        switch (propertyName) {
            default:
                return false;
        }
    }
    validateInstance(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceText) {
        const instanceElt = JSON.parse(instanceText);
        return this.validateInstanceElement(instanceElt);
    }
    validateInstanceElement(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt) {
        return this.validateInstanceInternal(instanceElt, undefined);
    }
    validateInstanceInternal(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        switch (this.dtdlVersion) {
            case 2: {
                return this.validateInstanceV2(instanceElt, instanceName);
            }
            case 3: {
                return this.validateInstanceV3(instanceElt, instanceName);
            }
        }
        return false;
    }
    validateInstanceV2(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        var _a, _b;
        if (!((_a = this.valueSchema) === null || _a === void 0 ? void 0 : _a.validateInstanceInternal(instanceElt, instanceName))) {
            return false;
        }
        if (!((_b = this.enumValues) === null || _b === void 0 ? void 0 : _b.some((val) => val.validateInstanceInternal(instanceElt, instanceName)))) {
            return false;
        }
        return true;
    }
    validateInstanceV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        var _a, _b;
        if (!((_a = this.valueSchema) === null || _a === void 0 ? void 0 : _a.validateInstanceInternal(instanceElt, instanceName))) {
            return false;
        }
        if (!((_b = this.enumValues) === null || _b === void 0 ? void 0 : _b.some((val) => val.validateInstanceInternal(instanceElt, instanceName)))) {
            return false;
        }
        return true;
    }
    /**
     * Set partition information.
     **/
    setPartitionInfo(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    partitionJsonText) {
        throw new Error(`attempt to set partition info on non-partition type EnumInfoInfo`);
    }
    applyTransformations(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this.dtdlVersion === 2) {
            this.applyTransformationsV2(model, parsingErrors);
        }
        if (this.dtdlVersion === 3) {
            this.applyTransformationsV3(model, parsingErrors);
        }
    }
    applyTransformationsV2(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) { }
    applyTransformationsV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) { }
    checkRestrictions(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this.dtdlVersion === 2) {
            this.checkRestrictionsV2(parsingErrors);
        }
        if (this.dtdlVersion === 3) {
            this.checkRestrictionsV3(parsingErrors);
        }
    }
    checkRestrictionsV2(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        var _a, _b, _c;
        const enumValuesNameSet = new Set();
        if (this.enumValues !== undefined) {
            for (const item of this.enumValues || []) {
                if (enumValuesNameSet.has(item.name)) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonUniquePropertyValue", {
                        cause: "{{primaryId:n}} property 'enumValues' contains more than one element whose property 'name' has value '{item.name}'.",
                        action: "Change the value of property  'name' to a string value that is unique across all values of 'enumValues'.",
                        primaryId: this.id,
                        property: "enumValues",
                        value: "name",
                    }));
                }
                enumValuesNameSet.add(item.name);
            }
        }
        const enumValuesEnumValueSet = new Set();
        if (this.enumValues !== undefined) {
            for (const item of this.enumValues || []) {
                if (enumValuesEnumValueSet.has(item.enumValue)) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonUniquePropertyValue", {
                        cause: "{{primaryId:n}} property 'enumValues' contains more than one element whose property 'enumValue' has value '{item.enumValue}'.",
                        action: "Change the value of property  'enumValue' to a string value that is unique across all values of 'enumValues'.",
                        primaryId: this.id,
                        property: "enumValues",
                        value: "enumValue",
                    }));
                }
                enumValuesEnumValueSet.add(item.enumValue);
            }
        }
        if (this.valueSchema !== undefined) {
            if (this.valueSchema.id !== "dtmi:dtdl:instance:Schema:integer;2" &&
                this.valueSchema.id !== "dtmi:dtdl:instance:Schema:string;2") {
                parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:incorrectPropertyValue", {
                    cause: "{{primaryId:n}} property 'valueSchema' has value {value} , but the value must be 'integer or string'.",
                    action: "Change the value of property  'valueSchema' to 'integer or string'.",
                    primaryId: this.id,
                    property: "valueSchema",
                    value: aggregateContext_1.AggregateContext.getTermOrUri(this.valueSchema.id),
                }));
            }
        }
        if (this._valueSchemaInstanceProperties !== undefined) {
            for (const instanceProp of this._valueSchemaInstanceProperties) {
                if (!((_a = this.valueSchema) === null || _a === void 0 ? void 0 : _a.validateInstanceElement(this.supplementalProperties[instanceProp]))) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonConformantPropertyValue", {
                        cause: "{{primaryId:n}} property '{{property}}' does not conform to the type specified by property 'valueSchema' ",
                        action: "Change the value of property '{{property}}' so that it conforms to property  'valueSchema'",
                        primaryId: this.id,
                        property: instanceProp,
                    }));
                }
            }
        }
        const tooDeepElementSchemaOrSchemaElementId = { ref: undefined };
        if (!this.checkDepthOfElementSchemaOrSchema(0, 5, tooDeepElementSchemaOrSchemaElementId, parsingErrors) &&
            tooDeepElementSchemaOrSchemaElementId !== undefined) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:excessiveDepth", {
                cause: `{primaryId:n} is at the root of a hierarchy that exceeds 5 levels -- {secondaryId:n} is at level 6.`,
                action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to reduce the nesting depth.`,
                primaryId: this.id,
                secondaryId: (_b = tooDeepElementSchemaOrSchemaElementId.ref) === null || _b === void 0 ? void 0 : _b.value,
            }));
        }
        const myUri = ((_c = this.valueSchema) === null || _c === void 0 ? void 0 : _c.id) || "";
        if (myUri === "dtmi:dtdl:instance:Schema:integer;2") {
            this.checkDescendantEnumValueDataType(new internalDtmi_1.InDTMI(this.id), "number", parsingErrors);
        }
        if (myUri === "dtmi:dtdl:instance:Schema:string;2") {
            this.checkDescendantEnumValueDataType(new internalDtmi_1.InDTMI(this.id), "string", parsingErrors);
        }
        if (myUri === "dtmi:dtdl:instance:Schema:boolean;2") {
            this.checkDescendantEnumValueDataType(new internalDtmi_1.InDTMI(this.id), "boolean", parsingErrors);
        }
    }
    checkRestrictionsV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        var _a, _b, _c;
        const enumValuesNameSet = new Set();
        if (this.enumValues !== undefined) {
            for (const item of this.enumValues || []) {
                if (enumValuesNameSet.has(item.name)) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonUniquePropertyValue", {
                        cause: "{{primaryId:n}} property 'enumValues' contains more than one element whose property 'name' has value '{item.name}'.",
                        action: "Change the value of property  'name' to a string value that is unique across all values of 'enumValues'.",
                        primaryId: this.id,
                        property: "enumValues",
                        value: "name",
                    }));
                }
                enumValuesNameSet.add(item.name);
            }
        }
        const enumValuesEnumValueSet = new Set();
        if (this.enumValues !== undefined) {
            for (const item of this.enumValues || []) {
                if (enumValuesEnumValueSet.has(item.enumValue)) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonUniquePropertyValue", {
                        cause: "{{primaryId:n}} property 'enumValues' contains more than one element whose property 'enumValue' has value '{item.enumValue}'.",
                        action: "Change the value of property  'enumValue' to a string value that is unique across all values of 'enumValues'.",
                        primaryId: this.id,
                        property: "enumValues",
                        value: "enumValue",
                    }));
                }
                enumValuesEnumValueSet.add(item.enumValue);
            }
        }
        if (this.valueSchema !== undefined) {
            if (this.valueSchema.id !== "dtmi:dtdl:instance:Schema:integer;3" &&
                this.valueSchema.id !== "dtmi:dtdl:instance:Schema:string;3") {
                parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:incorrectPropertyValue", {
                    cause: "{{primaryId:n}} property 'valueSchema' has value {value} , but the value must be 'integer or string'.",
                    action: "Change the value of property  'valueSchema' to 'integer or string'.",
                    primaryId: this.id,
                    property: "valueSchema",
                    value: aggregateContext_1.AggregateContext.getTermOrUri(this.valueSchema.id),
                }));
            }
        }
        if (this._valueSchemaInstanceProperties !== undefined) {
            for (const instanceProp of this._valueSchemaInstanceProperties) {
                if (!((_a = this.valueSchema) === null || _a === void 0 ? void 0 : _a.validateInstanceElement(this.supplementalProperties[instanceProp]))) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonConformantPropertyValue", {
                        cause: "{{primaryId:n}} property '{{property}}' does not conform to the type specified by property 'valueSchema' ",
                        action: "Change the value of property '{{property}}' so that it conforms to property  'valueSchema'",
                        primaryId: this.id,
                        property: instanceProp,
                    }));
                }
            }
        }
        const tooDeepElementSchemaOrSchemaElementId = { ref: undefined };
        if (!this.checkDepthOfElementSchemaOrSchema(0, 5, tooDeepElementSchemaOrSchemaElementId, parsingErrors) &&
            tooDeepElementSchemaOrSchemaElementId !== undefined) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:excessiveDepth", {
                cause: `{primaryId:n} is at the root of a hierarchy that exceeds 5 levels -- {secondaryId:n} is at level 6.`,
                action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to reduce the nesting depth.`,
                primaryId: this.id,
                secondaryId: (_b = tooDeepElementSchemaOrSchemaElementId.ref) === null || _b === void 0 ? void 0 : _b.value,
            }));
        }
        const myUri = ((_c = this.valueSchema) === null || _c === void 0 ? void 0 : _c.id) || "";
        if (myUri === "dtmi:dtdl:instance:Schema:integer;3") {
            this.checkDescendantEnumValueDataType(new internalDtmi_1.InDTMI(this.id), "number", parsingErrors);
        }
        if (myUri === "dtmi:dtdl:instance:Schema:string;3") {
            this.checkDescendantEnumValueDataType(new internalDtmi_1.InDTMI(this.id), "string", parsingErrors);
        }
        if (myUri === "dtmi:dtdl:instance:Schema:boolean;3") {
            this.checkDescendantEnumValueDataType(new internalDtmi_1.InDTMI(this.id), "boolean", parsingErrors);
        }
    }
    trySetObjectProperty(propertyName, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/explicit-module-boundary-types
    value, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key) {
        switch (propertyName) {
            case "enumValues":
            case "dtmi:dtdl:property:enumValues;2":
            case "dtmi:dtdl:property:enumValues;3":
                if (this.enumValues !== undefined) {
                    this.enumValues.push(value);
                    return true;
                }
                break;
            case "valueSchema":
            case "dtmi:dtdl:property:valueSchema;2":
            case "dtmi:dtdl:property:valueSchema;3":
                this.valueSchema = value;
                return true;
            default:
                break;
        }
        for (const supplementalType of this.supplementalTypes) {
            if (supplementalType.trySetObjectProperty(propertyName, value, key, this.supplementalProperties)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Check the nesting depth of all descendant elementSchema or schema properties.
     **/
    checkDepthOfElementSchemaOrSchema(depth, depthLimit, tooDeepElementId, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        var _a, _b;
        for (const item of this.enumValues || []) {
            if (!item.checkDepthOfElementSchemaOrSchema(depth, depthLimit, tooDeepElementId, parsingErrors)) {
                if (((_a = tooDeepElementId.ref) === null || _a === void 0 ? void 0 : _a.value) === this.id) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                        cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                        action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to remeve the recursion.`,
                        primaryId: this.id,
                    }));
                    tooDeepElementId.ref = undefined;
                }
                return false;
            }
        }
        if (this.valueSchema !== undefined) {
            if (!this.valueSchema.checkDepthOfElementSchemaOrSchema(depth, depthLimit, tooDeepElementId, parsingErrors)) {
                if (((_b = tooDeepElementId.ref) === null || _b === void 0 ? void 0 : _b.value) === this.id) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                        cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                        action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to remeve the recursion.`,
                        primaryId: this.id,
                    }));
                    tooDeepElementId.ref = undefined;
                }
                return false;
            }
        }
        tooDeepElementId.ref = undefined;
        return true;
    }
    tryGetDescendantSchemaOrContentsComponentNarrow(elementId) {
        if (this._checkedForDescendantSchemaOrContentsComponentNarrow) {
            elementId.ref = this._idOfDescendantSchemaOrContentsComponentNarrow;
            return this._idOfDescendantSchemaOrContentsComponentNarrow !== undefined;
        }
        this._checkedForDescendantSchemaOrContentsComponentNarrow = true;
        elementId.ref = undefined;
        return false;
    }
    checkDescendantEnumValueDataType(ancestorId, datatype, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this._checkedDescendantEnumValueDatatype !== datatype) {
            this._checkedDescendantEnumValueDatatype = datatype;
        }
        for (const item of this.enumValues || []) {
            item.checkDescendantEnumValueDataType(ancestorId, datatype, parsingErrors);
        }
        if (this.valueSchema !== undefined) {
            this.valueSchema.checkDescendantEnumValueDataType(ancestorId, datatype, parsingErrors);
        }
    }
    getTransitiveExtendsNarrow(depth, depthLimit, tooDeepElementId, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        const closure = new Set();
        tooDeepElementId.ref = undefined;
        return closure;
    }
    tryGetDescendantSchemaArray(elementId) {
        if (this._checkedForDescendantSchemaArray) {
            elementId.ref = this._idOfDescendantSchemaArray;
            return this._idOfDescendantSchemaArray !== undefined;
        }
        this._checkedForDescendantSchemaArray = true;
        for (const item of this.enumValues || []) {
            if (item.tryGetDescendantSchemaArray(elementId)) {
                this._idOfDescendantSchemaArray = elementId.ref;
                return true;
            }
        }
        if (this.valueSchema !== undefined) {
            if (this.valueSchema.tryGetDescendantSchemaArray(elementId)) {
                this._idOfDescendantSchemaArray = elementId.ref;
                return true;
            }
        }
        elementId.ref = undefined;
        return false;
    }
    getCountOfExtendsNarrow(parsingErrors) {
        if (this._countOfExtendsNarrowStatus === enum_1.TraversalStatus.Complete) {
            return this._countOfExtendsNarrowValue;
        }
        if (this._countOfExtendsNarrowStatus === enum_1.TraversalStatus.InProgress) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                action: `Change the value of one or more 'extends' properties in the hierarchy to remeve the recursion.`,
                primaryId: this.id,
            }));
            return 0;
        }
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.InProgress;
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.Complete;
        return this._countOfExtendsNarrowValue;
    }
    getCountOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrow(parsingErrors) {
        if (this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus ===
            enum_1.TraversalStatus.Complete) {
            return this
                ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue;
        }
        if (this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus ===
            enum_1.TraversalStatus.InProgress) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                action: `Change the value of one or more 'contents' or 'fields' or 'enumValues' or 'request' or 'response' or 'properties' or 'schema' or 'elementSchema' or 'mapValue' properties in the hierarchy to remeve the recursion.`,
                primaryId: this.id,
            }));
            return 0;
        }
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.InProgress;
        for (const item of this.enumValues || []) {
            this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue +=
                item.getCountOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrow(parsingErrors) + 1;
        }
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.Complete;
        return this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue;
    }
}
exports.EnumInfoImpl = EnumInfoImpl;
