// azure-iot-dtdl-parser/dist/parser/mapInfoImpl.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
/* eslint-disable valid-jsdoc */
/* eslint-disable guard-for-in */
/* eslint-disable no-empty */
/* eslint-disable no-empty-function */
/* eslint-disable no-unused-vars */
/* eslint-disable sort-imports */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MapInfoImpl = void 0;
const parserCollection_1 = require("./parserCollection");
const internalDtmi_1 = require("./internalDtmi");
const parsingErrorImpl_1 = require("./parsingErrorImpl");
const enum_1 = require("./enum");
class MapInfoImpl {
    constructor(dtdlVersion, id, childOf, definedIn, entityKind) {
        this._mapKeyValueConstraints = [];
        this._mapKeyInstanceProperties = [];
        this._mapKeyAllowedVersionsV2 = new Set().add(2);
        this._mapKeyAllowedVersionsV3 = new Set().add(3);
        this._mapValueValueConstraints = [];
        this._mapValueInstanceProperties = [];
        this._mapValueAllowedVersionsV2 = new Set().add(2);
        this._mapValueAllowedVersionsV3 = new Set().add(3);
        this.parserClass = parserCollection_1.ParserCollection.MapInfoParser;
        this.dtdlVersion = dtdlVersion;
        this.id = id;
        this.childOf = childOf;
        this.definedIn = definedIn;
        this.entityKind = entityKind;
        this.description = {};
        this.displayName = {};
        this.supplementalTypeIds = [];
        this.supplementalProperties = {};
        this.supplementalTypes = [];
        this.isPartition = false;
        this.undefinedTypes = [];
        this.undefinedProperties = {};
        this._checkedForDescendantSchemaOrContentsComponentNarrow = false;
        this._idOfDescendantSchemaOrContentsComponentNarrow = undefined;
        this._checkedDescendantEnumValueDatatype = undefined;
        this._checkedForDescendantSchemaArray = false;
        this._idOfDescendantSchemaArray = undefined;
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.NotStarted;
        this._countOfExtendsNarrowValue = 0;
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.NotStarted;
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue = 0;
    }
    static initialize() {
        this._versionlessTypes = new Set()
            .add("dtmi:dtdl:class:ComplexSchema")
            .add("dtmi:dtdl:class:Entity")
            .add("dtmi:dtdl:class:Map")
            .add("dtmi:dtdl:class:Schema");
    }
    addType(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    dtmi, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    supplementalType) {
        this.supplementalTypeIds.push(dtmi);
        if (supplementalType !== undefined) {
            this.supplementalTypes.push(supplementalType);
        }
        supplementalType.attachConstraints(this);
        supplementalType.bindInstanceProperties(this);
    }
    doesHaveType(typeId) {
        return (MapInfoImpl._versionlessTypes.has(new internalDtmi_1.InDTMI(typeId).versionless) ||
            this.supplementalTypes.some((st) => st.doesHaveType(typeId)));
    }
    addConstraint(propertyName, valueConstraint) {
        switch (propertyName) {
            case "mapKey":
                if (this._mapKeyValueConstraints === undefined) {
                    this._mapKeyValueConstraints = [];
                }
                this._mapKeyValueConstraints.push(valueConstraint);
                break;
            case "mapValue":
                if (this._mapValueValueConstraints === undefined) {
                    this._mapValueValueConstraints = [];
                }
                this._mapValueValueConstraints.push(valueConstraint);
                break;
        }
    }
    addInstanceProperty(propertyName, instancePropertyName) {
        switch (propertyName) {
            case "mapKey":
                if (this._mapKeyInstanceProperties === undefined) {
                    this._mapKeyInstanceProperties = [];
                }
                this._mapKeyInstanceProperties.push(instancePropertyName);
                break;
            case "mapValue":
                if (this._mapValueInstanceProperties === undefined) {
                    this._mapValueInstanceProperties = [];
                }
                this._mapValueInstanceProperties.push(instancePropertyName);
                break;
        }
    }
    doesPropertyDictContainKey(propertyName, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key) {
        switch (propertyName) {
            default:
                return false;
        }
    }
    validateInstance(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceText) {
        const instanceElt = JSON.parse(instanceText);
        return this.validateInstanceElement(instanceElt);
    }
    validateInstanceElement(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt) {
        return this.validateInstanceInternal(instanceElt, undefined);
    }
    validateInstanceInternal(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        switch (this.dtdlVersion) {
            case 2: {
                return this.validateInstanceV2(instanceElt, instanceName);
            }
            case 3: {
                return this.validateInstanceV3(instanceElt, instanceName);
            }
        }
        return false;
    }
    validateInstanceV2(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        var _a;
        if (typeof instanceElt !== "object") {
            return false;
        }
        // eslint-disable-next-line @typescript-eslint/ban-types
        for (const [key, value] of Object.entries(instanceElt)) {
            if (!((_a = this.mapValue) === null || _a === void 0 ? void 0 : _a.validateInstanceInternal(value, key))) {
                return false;
            }
        }
        return true;
    }
    validateInstanceV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        var _a;
        if (typeof instanceElt !== "object") {
            return false;
        }
        // eslint-disable-next-line @typescript-eslint/ban-types
        for (const [key, value] of Object.entries(instanceElt)) {
            if (!((_a = this.mapValue) === null || _a === void 0 ? void 0 : _a.validateInstanceInternal(value, key))) {
                return false;
            }
        }
        return true;
    }
    /**
     * Set partition information.
     **/
    setPartitionInfo(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    partitionJsonText) {
        throw new Error(`attempt to set partition info on non-partition type MapInfoInfo`);
    }
    applyTransformations(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this.dtdlVersion === 2) {
            this.applyTransformationsV2(model, parsingErrors);
        }
        if (this.dtdlVersion === 3) {
            this.applyTransformationsV3(model, parsingErrors);
        }
    }
    applyTransformationsV2(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) { }
    applyTransformationsV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) { }
    checkRestrictions(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this.dtdlVersion === 2) {
            this.checkRestrictionsV2(parsingErrors);
        }
        if (this.dtdlVersion === 3) {
            this.checkRestrictionsV3(parsingErrors);
        }
    }
    checkRestrictionsV2(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        var _a, _b, _c;
        if (this._mapKeyInstanceProperties !== undefined) {
            for (const instanceProp of this._mapKeyInstanceProperties) {
                if (!((_a = this.mapKey) === null || _a === void 0 ? void 0 : _a.validateInstanceElement(this.supplementalProperties[instanceProp]))) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonConformantPropertyValue", {
                        cause: "{{primaryId:n}} property '{{property}}' does not conform to the type specified by property 'mapKey' ",
                        action: "Change the value of property '{{property}}' so that it conforms to property  'mapKey'",
                        primaryId: this.id,
                        property: instanceProp,
                    }));
                }
            }
        }
        if (this._mapValueInstanceProperties !== undefined) {
            for (const instanceProp of this._mapValueInstanceProperties) {
                if (!((_b = this.mapValue) === null || _b === void 0 ? void 0 : _b.validateInstanceElement(this.supplementalProperties[instanceProp]))) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonConformantPropertyValue", {
                        cause: "{{primaryId:n}} property '{{property}}' does not conform to the type specified by property 'mapValue' ",
                        action: "Change the value of property '{{property}}' so that it conforms to property  'mapValue'",
                        primaryId: this.id,
                        property: instanceProp,
                    }));
                }
            }
        }
        const tooDeepElementSchemaOrSchemaElementId = { ref: undefined };
        if (!this.checkDepthOfElementSchemaOrSchema(0, 5, tooDeepElementSchemaOrSchemaElementId, parsingErrors) &&
            tooDeepElementSchemaOrSchemaElementId !== undefined) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:excessiveDepth", {
                cause: `{primaryId:n} is at the root of a hierarchy that exceeds 5 levels -- {secondaryId:n} is at level 6.`,
                action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to reduce the nesting depth.`,
                primaryId: this.id,
                secondaryId: (_c = tooDeepElementSchemaOrSchemaElementId.ref) === null || _c === void 0 ? void 0 : _c.value,
            }));
        }
    }
    checkRestrictionsV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        var _a, _b, _c;
        if (this._mapKeyInstanceProperties !== undefined) {
            for (const instanceProp of this._mapKeyInstanceProperties) {
                if (!((_a = this.mapKey) === null || _a === void 0 ? void 0 : _a.validateInstanceElement(this.supplementalProperties[instanceProp]))) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonConformantPropertyValue", {
                        cause: "{{primaryId:n}} property '{{property}}' does not conform to the type specified by property 'mapKey' ",
                        action: "Change the value of property '{{property}}' so that it conforms to property  'mapKey'",
                        primaryId: this.id,
                        property: instanceProp,
                    }));
                }
            }
        }
        if (this._mapValueInstanceProperties !== undefined) {
            for (const instanceProp of this._mapValueInstanceProperties) {
                if (!((_b = this.mapValue) === null || _b === void 0 ? void 0 : _b.validateInstanceElement(this.supplementalProperties[instanceProp]))) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:nonConformantPropertyValue", {
                        cause: "{{primaryId:n}} property '{{property}}' does not conform to the type specified by property 'mapValue' ",
                        action: "Change the value of property '{{property}}' so that it conforms to property  'mapValue'",
                        primaryId: this.id,
                        property: instanceProp,
                    }));
                }
            }
        }
        const tooDeepElementSchemaOrSchemaElementId = { ref: undefined };
        if (!this.checkDepthOfElementSchemaOrSchema(0, 5, tooDeepElementSchemaOrSchemaElementId, parsingErrors) &&
            tooDeepElementSchemaOrSchemaElementId !== undefined) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:excessiveDepth", {
                cause: `{primaryId:n} is at the root of a hierarchy that exceeds 5 levels -- {secondaryId:n} is at level 6.`,
                action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to reduce the nesting depth.`,
                primaryId: this.id,
                secondaryId: (_c = tooDeepElementSchemaOrSchemaElementId.ref) === null || _c === void 0 ? void 0 : _c.value,
            }));
        }
    }
    trySetObjectProperty(propertyName, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/explicit-module-boundary-types
    value, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key) {
        switch (propertyName) {
            case "mapKey":
            case "dtmi:dtdl:property:mapKey;2":
            case "dtmi:dtdl:property:mapKey;3":
                this.mapKey = value;
                return true;
            case "mapValue":
            case "dtmi:dtdl:property:mapValue;2":
            case "dtmi:dtdl:property:mapValue;3":
                this.mapValue = value;
                return true;
            default:
                break;
        }
        for (const supplementalType of this.supplementalTypes) {
            if (supplementalType.trySetObjectProperty(propertyName, value, key, this.supplementalProperties)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Check the nesting depth of all descendant elementSchema or schema properties.
     **/
    checkDepthOfElementSchemaOrSchema(depth, depthLimit, tooDeepElementId, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        var _a, _b;
        if (this.mapKey !== undefined) {
            if (!this.mapKey.checkDepthOfElementSchemaOrSchema(depth, depthLimit, tooDeepElementId, parsingErrors)) {
                if (((_a = tooDeepElementId.ref) === null || _a === void 0 ? void 0 : _a.value) === this.id) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                        cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                        action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to remeve the recursion.`,
                        primaryId: this.id,
                    }));
                    tooDeepElementId.ref = undefined;
                }
                return false;
            }
        }
        if (this.mapValue !== undefined) {
            if (!this.mapValue.checkDepthOfElementSchemaOrSchema(depth, depthLimit, tooDeepElementId, parsingErrors)) {
                if (((_b = tooDeepElementId.ref) === null || _b === void 0 ? void 0 : _b.value) === this.id) {
                    parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                        cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                        action: `Change the value of one or more 'elementSchema' or 'schema' properties in the hierarchy to remeve the recursion.`,
                        primaryId: this.id,
                    }));
                    tooDeepElementId.ref = undefined;
                }
                return false;
            }
        }
        tooDeepElementId.ref = undefined;
        return true;
    }
    tryGetDescendantSchemaOrContentsComponentNarrow(elementId) {
        if (this._checkedForDescendantSchemaOrContentsComponentNarrow) {
            elementId.ref = this._idOfDescendantSchemaOrContentsComponentNarrow;
            return this._idOfDescendantSchemaOrContentsComponentNarrow !== undefined;
        }
        this._checkedForDescendantSchemaOrContentsComponentNarrow = true;
        elementId.ref = undefined;
        return false;
    }
    checkDescendantEnumValueDataType(ancestorId, datatype, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this._checkedDescendantEnumValueDatatype !== datatype) {
            this._checkedDescendantEnumValueDatatype = datatype;
        }
        if (this.mapKey !== undefined) {
            this.mapKey.checkDescendantEnumValueDataType(ancestorId, datatype, parsingErrors);
        }
        if (this.mapValue !== undefined) {
            this.mapValue.checkDescendantEnumValueDataType(ancestorId, datatype, parsingErrors);
        }
    }
    getTransitiveExtendsNarrow(depth, depthLimit, tooDeepElementId, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        const closure = new Set();
        tooDeepElementId.ref = undefined;
        return closure;
    }
    tryGetDescendantSchemaArray(elementId) {
        if (this._checkedForDescendantSchemaArray) {
            elementId.ref = this._idOfDescendantSchemaArray;
            return this._idOfDescendantSchemaArray !== undefined;
        }
        this._checkedForDescendantSchemaArray = true;
        if (this.mapKey !== undefined) {
            if (this.mapKey.tryGetDescendantSchemaArray(elementId)) {
                this._idOfDescendantSchemaArray = elementId.ref;
                return true;
            }
        }
        if (this.mapValue !== undefined) {
            if (this.mapValue.tryGetDescendantSchemaArray(elementId)) {
                this._idOfDescendantSchemaArray = elementId.ref;
                return true;
            }
        }
        elementId.ref = undefined;
        return false;
    }
    getCountOfExtendsNarrow(parsingErrors) {
        if (this._countOfExtendsNarrowStatus === enum_1.TraversalStatus.Complete) {
            return this._countOfExtendsNarrowValue;
        }
        if (this._countOfExtendsNarrowStatus === enum_1.TraversalStatus.InProgress) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                action: `Change the value of one or more 'extends' properties in the hierarchy to remeve the recursion.`,
                primaryId: this.id,
            }));
            return 0;
        }
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.InProgress;
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.Complete;
        return this._countOfExtendsNarrowValue;
    }
    getCountOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrow(parsingErrors) {
        if (this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus ===
            enum_1.TraversalStatus.Complete) {
            return this
                ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue;
        }
        if (this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus ===
            enum_1.TraversalStatus.InProgress) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                action: `Change the value of one or more 'contents' or 'fields' or 'enumValues' or 'request' or 'response' or 'properties' or 'schema' or 'elementSchema' or 'mapValue' properties in the hierarchy to remeve the recursion.`,
                primaryId: this.id,
            }));
            return 0;
        }
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.InProgress;
        if (this.mapValue !== undefined) {
            this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue +=
                this.mapValue.getCountOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrow(parsingErrors) + 1;
        }
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.Complete;
        return this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue;
    }
}
exports.MapInfoImpl = MapInfoImpl;
