// azure-iot-dtdl-parser/dist/parser/namedLatentTypeInfoImpl.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
/* eslint-disable valid-jsdoc */
/* eslint-disable guard-for-in */
/* eslint-disable no-empty */
/* eslint-disable no-empty-function */
/* eslint-disable no-unused-vars */
/* eslint-disable sort-imports */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamedLatentTypeInfoImpl = void 0;
const parserCollection_1 = require("./parserCollection");
const internalDtmi_1 = require("./internalDtmi");
const enum_1 = require("./enum");
const parsingErrorImpl_1 = require("./parsingErrorImpl");
class NamedLatentTypeInfoImpl {
    constructor(dtdlVersion, id, childOf, definedIn, entityKind) {
        this.namePropertyRegexPatternV3 = /^[A-Za-z](?:[A-Za-z0-9_]*[A-Za-z0-9])?$/;
        this.parserClass = parserCollection_1.ParserCollection.NamedLatentTypeInfoParser;
        this.dtdlVersion = dtdlVersion;
        this.id = id;
        this.childOf = childOf;
        this.definedIn = definedIn;
        this.entityKind = entityKind;
        this.description = {};
        this.displayName = {};
        this.supplementalTypeIds = [];
        this.supplementalProperties = {};
        this.supplementalTypes = [];
        this.isPartition = false;
        this.undefinedTypes = [];
        this.undefinedProperties = {};
        this._checkedForDescendantSchemaOrContentsComponentNarrow = false;
        this._idOfDescendantSchemaOrContentsComponentNarrow = undefined;
        this._checkedDescendantEnumValueDatatype = undefined;
        this._checkedForDescendantSchemaArray = false;
        this._idOfDescendantSchemaArray = undefined;
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.NotStarted;
        this._countOfExtendsNarrowValue = 0;
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.NotStarted;
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue = 0;
    }
    static initialize() {
        this._versionlessTypes = new Set()
            .add("dtmi:dtdl:class:Entity")
            .add("dtmi:dtdl:class:NamedEntity")
            .add("dtmi:dtdl:class:NamedLatentType");
    }
    addType(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    dtmi, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    supplementalType) {
        this.supplementalTypeIds.push(dtmi);
        if (supplementalType !== undefined) {
            this.supplementalTypes.push(supplementalType);
        }
    }
    doesHaveType(typeId) {
        return (NamedLatentTypeInfoImpl._versionlessTypes.has(new internalDtmi_1.InDTMI(typeId).versionless) ||
            this.supplementalTypes.some((st) => st.doesHaveType(typeId)));
    }
    doesPropertyDictContainKey(propertyName, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key) {
        switch (propertyName) {
            default:
                return false;
        }
    }
    validateInstance(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceText) {
        const instanceElt = JSON.parse(instanceText);
        return this.validateInstanceElement(instanceElt);
    }
    validateInstanceElement(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt) {
        return false;
    }
    validateInstanceInternal(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        return false;
    }
    validateInstanceV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceElt, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    instanceName) {
        return false;
    }
    /**
     * Set partition information.
     **/
    setPartitionInfo(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    partitionJsonText) {
        throw new Error(`attempt to set partition info on non-partition type NamedLatentTypeInfoInfo`);
    }
    applyTransformations(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this.dtdlVersion === 3) {
            this.applyTransformationsV3(model, parsingErrors);
        }
    }
    applyTransformationsV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    model, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) { }
    checkRestrictions(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this.dtdlVersion === 3) {
            this.checkRestrictionsV3(parsingErrors);
        }
    }
    checkRestrictionsV3(
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) { }
    trySetObjectProperty(propertyName, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/explicit-module-boundary-types
    value, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    key) {
        switch (propertyName) {
            default:
                break;
        }
        for (const supplementalType of this.supplementalTypes) {
            if (supplementalType.trySetObjectProperty(propertyName, value, key, this.supplementalProperties)) {
                return true;
            }
        }
        return false;
    }
    /**
     * Check the nesting depth of all descendant elementSchema or schema properties.
     **/
    checkDepthOfElementSchemaOrSchema(depth, depthLimit, tooDeepElementId, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        tooDeepElementId.ref = undefined;
        return true;
    }
    tryGetDescendantSchemaOrContentsComponentNarrow(elementId) {
        if (this._checkedForDescendantSchemaOrContentsComponentNarrow) {
            elementId.ref = this._idOfDescendantSchemaOrContentsComponentNarrow;
            return this._idOfDescendantSchemaOrContentsComponentNarrow !== undefined;
        }
        this._checkedForDescendantSchemaOrContentsComponentNarrow = true;
        elementId.ref = undefined;
        return false;
    }
    checkDescendantEnumValueDataType(ancestorId, datatype, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        if (this._checkedDescendantEnumValueDatatype !== datatype) {
            this._checkedDescendantEnumValueDatatype = datatype;
        }
    }
    getTransitiveExtendsNarrow(depth, depthLimit, tooDeepElementId, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    parsingErrors) {
        const closure = new Set();
        tooDeepElementId.ref = undefined;
        return closure;
    }
    tryGetDescendantSchemaArray(elementId) {
        if (this._checkedForDescendantSchemaArray) {
            elementId.ref = this._idOfDescendantSchemaArray;
            return this._idOfDescendantSchemaArray !== undefined;
        }
        this._checkedForDescendantSchemaArray = true;
        elementId.ref = undefined;
        return false;
    }
    getCountOfExtendsNarrow(parsingErrors) {
        if (this._countOfExtendsNarrowStatus === enum_1.TraversalStatus.Complete) {
            return this._countOfExtendsNarrowValue;
        }
        if (this._countOfExtendsNarrowStatus === enum_1.TraversalStatus.InProgress) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                action: `Change the value of one or more 'extends' properties in the hierarchy to remeve the recursion.`,
                primaryId: this.id,
            }));
            return 0;
        }
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.InProgress;
        this._countOfExtendsNarrowStatus = enum_1.TraversalStatus.Complete;
        return this._countOfExtendsNarrowValue;
    }
    getCountOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrow(parsingErrors) {
        if (this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus ===
            enum_1.TraversalStatus.Complete) {
            return this
                ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue;
        }
        if (this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus ===
            enum_1.TraversalStatus.InProgress) {
            parsingErrors.push((0, parsingErrorImpl_1.createParsingError)("dtmi:dtdl:parsingError:recursiveStructure", {
                cause: `{primaryId:n} is at the root of a hierarchy that includes itself.`,
                action: `Change the value of one or more 'contents' or 'fields' or 'enumValues' or 'request' or 'response' or 'properties' or 'schema' or 'elementSchema' or 'mapValue' properties in the hierarchy to remeve the recursion.`,
                primaryId: this.id,
            }));
            return 0;
        }
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.InProgress;
        this._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowStatus =
            enum_1.TraversalStatus.Complete;
        return this
            ._countOfContentsOrFieldsOrEnumValuesOrRequestOrResponseOrPropertiesOrSchemaOrElementSchemaOrMapValueNarrowValue;
    }
}
exports.NamedLatentTypeInfoImpl = NamedLatentTypeInfoImpl;
